core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala

```diff
@@ -243,6 +248,87 @@ class SnappySessionState(snappySession: SnappySession)
     }
   }
 
+  /**
+    * Orders the join keys as per the  underlying partitioning keys ordering of the table.
+    */
+  object OrderJoinConditions extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+      case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, otherCondition, left, right) =>
+        prepareOrderedCondition(joinType, left, right, leftKeys, rightKeys, otherCondition)
+    }
+
+    @tailrec def unAlias(e: Expression): Expression = e match {
+      case a: Alias => unAlias(a.child)
+      case _ => e
+    }
+
+    def getKeyOrder(plan: LogicalPlan, joinKeys: Seq[Expression],
+        partitioning: Seq[NamedExpression]): Seq[Int] = {
+      val part = partitioning.map(unAlias)
+      lazy val planExpressions = plan.expressions
+      val keyOrder = joinKeys.map { k =>
+        val key = unAlias(k)
+        val i = part.indexWhere(_.semanticEquals(key))
+        if (i < 0) {
+          // search for any view aliases (SNAP-2204)
+          key match {
+            case ke: NamedExpression =>
+              planExpressions.collectFirst {
+                case a: Alias if ke.exprId == a.exprId => unAlias(a.child)
+                case e: NamedExpression if (ke ne e) && ke.exprId == e.exprId => e
+              } match {
+                case Some(e) => part.indexWhere(_.semanticEquals(e))
+                case None => Int.MaxValue
+              }
+            case _ => Int.MaxValue
+          }
+        } else i
+      }
+      keyOrder
+    }
+
+    def getPartCols(plan: LogicalPlan): Seq[NamedExpression] = {
+      plan match {
+        case PhysicalScan(_, _, child) => child match {
+          case r@LogicalRelation(scan: PartitionedDataSourceScan, _, _) =>
+            // send back numPartitions=1 for replicated table since collocated
+            if (!scan.isPartitioned) return Nil
+            val partCols = scan.partitionColumns.map(colName =>
+              r.resolveQuoted(colName, analysis.caseInsensitiveResolution)
+                  .getOrElse(throw new AnalysisException(
+                    s"""Cannot resolve column "$colName" among (${r.output})""")))
+            partCols
+          case _ => Nil
+        }
+        case _ => Nil
+      }
+    }
+
```

Comment:
Above two methods look similar to those in HashJoinStrategies (apart from Nil vs Int.MaxValue difference). Can be refactored into a base trait.

Or perhaps once this is applied, then that code in HashJoinStrategies can be removed and substituted by a simpler comparison (or this can determine colocated join here itself and inject a separate "ColocatedJoin" plan which can be resolved appropriately in HashJoinStrategies).

Reply:
Refactored into a single trait.